### DISCUSSION TITLE
How to refresh with writableAtom with LocalStorage + Do we have a way to await Jotai to read and wait the value  from localStorage?

### DISCUSSION THREAD

**Discussion by meotimdihia** [pmndrs/jotai#2156] [Q&A] [answered]
Hi, I have 2 questions: 
## How to refresh with writableAtom with LocalStorage
please help me with this. I can't do it to work. This is the demo:

https://stackblitz.com/edit/jotai-indexeddb-persist-odx8hg?file=index.tsx,package.json

```ts
export const zoomStorageAtom = atomWithStorage<{ [key: number]: number }>(
  'zoomr',
  { 1: 100 },
  //@ts-ignore
  defaultStorage
);

export const zoomAtom = atom(
  (get) => {
    return get(zoomStorageAtom);
  },
  (get, set, newValue: { id: number; value: number }) => {
    const zoomAtom = get(zoomStorageAtom);
    zoomAtom[newValue.id] = newValue.value;
    set(zoomStorageAtom, zoomAtom);
  }
);
```
## Do we have a way to await Jotai to read and wait the value  from localStorage?
 
in this example, it looks like Jotai returns the default value and re-renders when it gets the value from LocalStorage

https://github.com/pmndrs/jotai/assets/300961/9a1c8aca-98b0-4edb-81d1-7af3c7ec03c5

**Comment by dai-shi** [comment, marked as answer]
> How to refresh with writableAtom with LocalStorage

You must update the value immutably:
```js
export const zoomAtom = atom(
  (get) => {
    return get(zoomStorageAtom);
  },
  (get, set, newValue: { id: number; value: number }) => {
    const current = get(zoomStorageAtom);
    const next = { ...current, ...newValue };
    set(zoomStorageAtom, next);
  }
);
```

**Comment by dai-shi** [comment]
> Do we have a way to await Jotai to read and wait the value from localStorage?

We have `unstable_getOnInit` option. Please try it and see if it works as you expect.

**Reply by meotimdihia** [reply]
Thanks for answering me. It fixed the problem:

https://stackblitz.com/edit/jotai-indexeddb-persist-zasjbn?file=index.tsx,package.json
